specification/dartLangSpec.tex

```diff
@@ -16484,160 +16766,282 @@ \subsection{Exports}
 }
 
+\commentary{%
+Note that it is possible to use \HIDE{} or \SHOW{} on an identifier
+which is not in the given namespace.%
+}
+\rationale{%
+Allowing this prevents situations where, e.g.,
+removing a declaration from a library $L$ would cause
+breakage in a library that imports $L$.%
+}
+
+
+\subsection{Conflict Merging of Namespaces}
+\LMLabel{conflictMergingOfNamespaces}
+
+\LMHash{}%
+In this section we define an operation on namespaces
+which eliminates certain bindings of names in case of name clashes,
+and keeps track of such name clashes using a special value,
+\Index{\ConflictValue}.
+
+\LMHash{}%
+When a name $n$ is mapped to \ConflictValue{} by a namespace,
+we say that $n$ is
+\IndexCustom{conflicted}{conflicted name}.
+A lookup for a conflicted name will succeed according to
+the normal rules for namespaces and lexical scoping,
+but it is a compile-time error at the location where the name is used
+that this name is conflicted.
+
 \LMHash{}%
-The second step starts from \NSN{i}
-and obtains the final result, \Namespace{i},
-by eliminating each key $n$, bound to declaration $D$,
-where one of the following conditions hold:
+Let \List{\metavar{NS}}{1}{m} be a list of namespaces
+(\commentary{the list may or may not contain duplicates}).
+The
+\IndexCustom{conflict merging}{namespace!conflict merging}
+of \List{\metavar{NS}}{1}{m} is then
+$\NamespaceName{\metavar{conflict}} \cup
+\NamespaceName{\metavar{narrowed},1} \cup\,
+\ldots\ \NamespaceName{\metavar{narrowed},m}$,
+where the namespaces \NamespaceName{\metavar{conflict}} and
+\NamespaceName{\metavar{narrowed},i}, $i \in 1 .. m$,
+are specified as follows:
+
+\LMHash{}%
+\NamespaceName{\metavar{conflict}} is empty except for the
+bindings mentioned below.
+For each $i \in 1 .. m$,
+\NamespaceName{\metavar{narrowed},i} is identical to \NamespaceName{i},
+except that the former is undefined at each name $n$ where
+the latter is defined,
+and one of the following conditions holds:
 
 \begin{itemize}
-\item \Namespace{L} maps $n$ to a declaration.
+\item There exists a $j \in 1 .. m$ and name $n'$ such that
+  \NamespaceName{j} is defined at $n'$,
+  $n$ and $n'$ have the same basename,
+  \Namespace{i}{n} is a declaration in a system library,
+  and \Namespace{j}{n'} is a declaration in a non-system library.
   \commentary{%
-    In other words, a declaration in $L$ shadows
-    re-exported declarations with the same name.%
+    So a declaration from a non-system library shadows
+    declarations from system libraries.%
   }
-\item The first case does not apply,
-  $L_i$ is a system library,
-  there exists a $j \not= i$ in $1 .. m$ such that
-  $L_j$ is not a system library,
-  and \NSN{j} maps $n$ to a declaration.
+\item Otherwise, there exists a $j \in 1 .. m$ and name $n'$ such that
+  \NamespaceName{j} is defined at $n'$,
+  $n$ and $n'$ have the same basename,
+  \Namespace{i}{n} and \Namespace{j}{n'} are not the same declaration
```

Comment:
They can't be declarations (because there are no conflicts involving an import prefix and a top-level declaration). I added commentary to make that point.

Reply:
There are *potential* conflicts between prefix imports. You have multiple NS_one,i's which contain the same name (mapped to the same namespace). Or, as this is written, it's possible to conflict with itself, because you don't require *i* and *j'* to be distinct.
The text below is the text which says when there is *not* a conflict, and it only works for declarations.

So, for NS_i and which is a prefix namespace mapping `n` to `NSp`,
there exists a *j* (*j* = *i*) and *n'* (*n'* = *n*) such that NS_j is defined at *n'*, *n'* and *n* have the same basename, and NSp and NSp is *not* the same declaration ... because it's not a declaration at all.

The way this is phrased, you *need* to either not mention "declaration" or to say "or same namespace".

Reply:
Sorry, I just saw this comment now. I'll handle changes as needed in a new PR.

> So, for NS_i and which is a prefix namespace mapping n to NSp,
> there exists a j (j = i) and n' (n' = n) such that NS_j is defined at
> n', n' and n have the same basename, and NSp and NSp is not the
> same declaration ... because it's not a declaration at all.

True, but pushing hard on a slight ambiguity. ;-)

PR https://github.com/dart-lang/language/pull/1259 makes an adjustment to avoid that ambiguity.